apis/python/src/tiledbsoma/io/registration/signatures.py

```diff
@@ -0,0 +1,305 @@
+            if "varm" in exp.ms[measurement_name]:
+                for varm_layer_name in exp.ms[measurement_name].varm.keys():
+                    varm = exp.ms[measurement_name].varm[varm_layer_name]
+                    varm_dtypes[varm_layer_name] = str(
+                        varm.schema.field("soma_data").type
+                    )
+
+            return cls(
+                obs_schema=obs_schema,
+                var_schema=var_schema,
+                X_dtypes=X_dtypes,
+                raw_X_dtype=raw_X_dtype,
+                raw_var_schema=raw_var_schema,
+                obsm_dtypes=obsm_dtypes,
+                varm_dtypes=varm_dtypes,
+            )
+
+    @classmethod
+    def compatible(cls, signatures: Dict[str, Self]) -> Tuple[bool, Optional[str]]:
+        """
+        Determines if two signatures from SOMA experiment or AnnData/H5AD will be safe from
+        schema-incompatibility at ingestion time. On success, the second argument is None; on
+        failure, the second argument can be used for reporting to the user.
+        """
+        if len(signatures) < 2:
+            return (True, None)
+        names = list(signatures.keys())
+        first_name = names[0]
+        for other_name in names[1:]:
+            siga = signatures[first_name]
+            sigb = signatures[other_name]
+            if not siga.compatibleWith(sigb):
+                msg = f"Incompatible signatures {first_name!r}, {other_name!r}:\n{siga.toJSON()}\n{sigb.toJSON()}"
+                return (False, msg)
+        return (True, None)
+
+    def compatibleWith(self, other: Self) -> bool:
+        """
+        Pairwise helper method for ``compatible``. Reasons for incompatibility are currently advised
+        to be handled a level up by simply showing the user the failed signature pair.
+        """
+
+        if self.obs_schema != other.obs_schema:
+            return False
+
+        if self.var_schema != other.var_schema:
+            return False
+
+        if self.X_dtypes != other.X_dtypes:
+            return False
+
+        if self.raw_X_dtype != other.raw_X_dtype:
+            return False
+        if self.raw_var_schema != other.raw_var_schema:
+            return False
+
+        if self.obsm_dtypes != other.obsm_dtypes:
+            return False
+        if self.varm_dtypes != other.varm_dtypes:
+            return False
+
+        return True
+
+    def toJSON(self) -> str:
+        """Presents a signature as JSON which is suitable for distributed logging."""
+        return json.dumps(self, default=lambda o: o.__dict__, sort_keys=True, indent=4)
+
+    @classmethod
+    def fromJSON(cls, s: str) -> Self:
+        dikt = json.loads(s)
+        return cls(
+            dikt["obs_schema"],
+            dikt["var_schema"],
+            dikt["raw_var_schema"],
+            dikt["X_dtypes"],
+            dikt["raw_X_dtype"],
+            dikt["obsm_dtypes"],
+            dikt["varm_dtypes"],
```

Comment:
You can save some typing with `cls(**dikt)`. Also maybe worth declaring the class as `kw_only=True`, because it seems like there’s no obviously correct order for these.